### DISCUSSION TITLE
some files are emptied or truncated

### DISCUSSION THREAD

**Discussion by jererc** [abraunegg/onedrive#2953] [Q&A]
Hello!

I run onedrive on ubuntu 2404, v2.5.2-1, as a service.
I found out some local files (text) are empty, and just found one file has been truncated.
The remote files are ok (the content is what I expect).
The files can be updated by other clients (win11 or ubuntu hosts or vms) for the same account.

I stopped the service and ran onedrive with the --resync arg.
It downloaded some files, and uploaded some old files (which had been removed days ago), and did not recover the truncated or empty files.

How can I debug this?

**Comment by abraunegg** [comment]
@jererc 

The first question I have is which system is used to create|modify the files?

> It downloaded some files, and uploaded some old files (which had been removed days ago), and did not recover the truncated or empty files.

This would mean that the files are still local which is why they were uploaded

> How can I debug this?

Firstly, by generating a debug log to understand what the client is doing - please read https://github.com/abraunegg/onedrive/wiki/Generate-debug-log-for-support

The next aspect is to understand where are you making file changes.

If nothing is meant to change on the Ubuntu platform, I would use `--download-only` and include the `--cleanup-local-files` option so that the Ubuntu platform(s) are only downloading changed files.

Secondly, disable any/all file indexing on Ubuntu .. as this process , depending on what you are using , will *modify* your files in the background to add their metadata to track it has been indexed .. which, then triggers the client to upload the file (because it was modified). Some file indexers also delete & re-create the file ... which, could also account for your experience.

Understanding *how* your system is setup and how you are modifying your files and comparing this to what you are experiencing is key. A diagram to illustrate what you are expecting and/or wanting this to work would go a long way.

**Comment by jererc** [comment]
The ubuntu vms are updating the files, but one was updated by a win11 editor lately so I'm not sure of anything, any host/vm can update.

First, I'll run the service with the --verbose flags, and think about disabling the tracking once I understand better.

Thank you for the info!

**Comment by jererc** [comment]
> Secondly, disable any/all file indexing on Ubuntu .. as this process , depending on what you are using , will modify your files in the background to add their metadata to track it has been indexed .. which, then triggers the client to upload the file (because it was modified). Some file indexers also delete & re-create the file ... which, could also account for your experience.

It seems the ubuntu tracker can modify files meta (write back feature - https://wiki.gnome.org/Attic/Tracker/Documentation/Writeback) but it's supposed to be disabled by default and I could not find any user config overwriting it.
I also checked my code editor (sublimetext) indexing and it does not update files, it only monitors them.

I'm testing megasync now, we'll see what happens.

**Reply by abraunegg** [reply]
How does your code editor write files when it updates them?

For example - WPS Office - last time I looked at WPS Office - regardless of content change, deletes the original file and writes a whole new file.

If 'onedrive' is monitoring that file, an inotify delete event is raised by the platform for the delete, before a new file is creates.

Does your code editor do something similar.

**Reply by jererc** [reply]
I won't read their codebase, I read enough code every day ;)

chatgpt is telling me a good story about atomic renaming:

> Once the temporary file is written successfully, Sublime Text renames it to replace the original file. Renaming is an atomic operation on most file systems, meaning it happens instantly, reducing the risk of file corruption.
This approach ensures that the original file remains intact until the new version is fully written, which is especially useful if there’s a power loss or other interruption during the save process.

when asking for sources though...

> Sublime Text’s official documentation is relatively sparse and does not directly detail the save process. However, users on the Sublime Text Forum and the unofficial Sublime Text Documentation discuss the editor’s behavior, noting that atomic saves (via a temporary file) are the default method to avoid file corruption.

and blah blah common practices

> Many modern text editors, such as Visual Studio Code and Atom, use this same atomic-saving approach to minimize the risk of data loss. Sublime Text appears to follow this common practice, as evidenced by the temporary files that appear briefly in the directory during saves.

**Reply by jererc** [reply]
but I also remember having issues with files outside the code editor's scope, some 0 byte binary files (windows .lnk files) in a backup dir

**Reply by abraunegg** [reply]
Zero byte files are handled by this 'onedrive' client without issue ... Windows client not so much

**Reply by jererc** [reply]
the lnk files were ok on windows and remotely, they were 0 byte size locally on the ubuntu 2404 vm

**Reply by jererc** [reply]
frankly, I suspect an issue in the code with poor network condition (5G in mauritius... I can get minor disconnections from time to time and latency issues)
something like the download is said ok, the row is saved in the db but the actual content is fucked

**Reply by abraunegg** [reply]
The application performs integrity checks of all files it downloads and uploads by default - please read the architecture documentation as your reference

**Reply by abraunegg** [reply]
As you are also using Ubuntu (why ... your choice, but there are much better distributions out there) please ensure you read this to ensure you configure the client correctly for the *broken* curl version your distribution provides:

https://github.com/abraunegg/onedrive/blob/master/docs/usage.md#compatibility-with-curl

**Comment by jererc** [comment]
damned!
thanks for the doc

but why letting the service run with a shitty dependency? 
I did not see even a warning in the logs, I'd even expect to see an error

**Reply by jererc** [reply]
I went through your doc (I should have done that first, I'm too lazy), and with all the limitations then I guess I can forget microsoft OneDrive for what I intend to do.
This is very interesting though

**Reply by abraunegg** [reply]
In all versions less than v2.5.3 (not a typo) there is no automatic configuration and/or warning printed.

This was added in PR's post v2.5.2 release and the current documentation is reflwctive of 'master' codebase.
